R/bookmark-state.R

```diff
@@ -0,0 +1,960 @@
+
+            if (!is.null(qsValues[["__subapp__"]]) && qsValues[["__subapp__"]] == 1) {
+              # Ignore subapps in shiny docs
+              self$reset()
+
+            } else if (!is.null(qsValues[["_state_id_"]]) && nzchar(qsValues[["_state_id_"]])) {
+              # If we have a "_state_id_" key, restore from saved state and
+              # ignore other key/value pairs. If not, restore from key/value
+              # pairs in the query string.
+              self$active <- TRUE
+              private$loadStateQueryString(queryString)
+
+            } else {
+              # The query string contains the saved keys and values
+              self$active <- TRUE
+              private$decodeStateQueryString(queryString)
+            }
+          }),
+          error = function(e) {
+            # If there's an error in restoring problem, just reset these values
+            self$reset()
+            self$initErrorMessage <- e$message
+            warning(e$message)
+          }
+        )
+      }
+    },
+
+    reset = function() {
+      self$active <- FALSE
+      self$initErrorMessage <- NULL
+      self$input <- RestoreInputSet$new(list())
+      self$values <- new.env(parent = emptyenv())
+      self$dir <- NULL
+    },
+
+    # This should be called before a restore context is popped off the stack.
+    flushPending = function() {
+      self$input$flushPending()
+    },
+
+
+    # Returns a list representation of the RestoreContext object. This is passed
+    # to the app author's onRestore function. An important difference between
+    # the RestoreContext object and the list is that the former's `input` field
+    # is a RestoreInputSet object, while the latter's `input` field is just a
+    # list.
+    asList = function() {
+      list(
+        input = self$input$asList(),
+        dir = self$dir,
+        values = self$values
+      )
+    }
+  ),
+
+  private = list(
+    # Given a query string with a _state_id_, load saved state with that ID.
+    loadStateQueryString = function(queryString) {
+      values <- parseQueryString(queryString, nested = TRUE)
+      id <- values[["_state_id_"]]
+
+      # This function is passed to the loadInterface function; given a
+      # directory, it will load state from that directory
+      loadFun <- function(stateDir) {
+        self$dir <- stateDir
+
+        inputValues <- readRDS(file.path(stateDir, "input.rds"))
+        self$input <- RestoreInputSet$new(inputValues)
+
+        valuesFile <- file.path(stateDir, "values.rds")
+        if (file.exists(valuesFile)) {
+          self$values <- readRDS(valuesFile)
+        }
+      }
+
+      loadInterface <- getShinyOption("load.interface", default = loadInterfaceLocal)
+      loadInterface(id, loadFun)
+
+      invisible()
+    },
+
+    # Given a query string with values encoded in it, restore saved state
+    # from those values.
+    decodeStateQueryString = function(queryString) {
```

Comment:
There are a bunch of unit tests here:
https://github.com/wch/shiny/blob/bookmarkable-state/tests/testthat/test-bookmarking.R

`_inputs_` and `_values_` are both optional. Non-bookmark params can occur only before bookmark ones. That could be changed, though.